### ISSUE TITLE
create and release a Docker image to run validations

### ISSUE THREAD

**Issue by elray1** [Infectious-Disease-Modeling-Hubs/hubValidations#57]
this would likely live in a different repo, but am filing the issue here for now

**Comment by annakrystalli** [comment]
Here's a vignette on the workflow: https://65a1509bed4bd603ae77b0c2--hubverse-pkgdown-preview.netlify.app/articles/validate-submission

**Comment by nickreich** [comment]
I removed the assignees from this issue, as I don't think this is being actively worked on.

**Comment by zkamvar** [comment]
Initial container in https://github.com/hubverse-org/test-docker-hubValidations. I will follow up tomorrow by experimenting with incorporating this into a fork of a hub.

**Comment by zkamvar** [comment]
I've tested it in https://github.com/zkamvar/variant-nowcast-hub/pull/1, which was created by duplicating the output in https://github.com/reichlab/variant-nowcast-hub/pull/246 and adding an error to the value column


Some observations:

1. The test runs about 2 minutes faster than [the original run](https://github.com/reichlab/variant-nowcast-hub/actions/runs/12562879972/job/35023947555) because we did not have to install R or check for dependency updates.
2. The workflow is shorter
3. The container takes 15 to 20 minutes to build on GitHub actions, but once it's built, it only takes 20 to 30 seconds to provision in a GitHub workflow. 
4. I do not yet know how to version this, but I can dig into it

**Comment by annakrystalli** [comment]
My feeling is that we should not be wrapping the validation code within scripts in the container. I feel using an [image that contains all necessary software](https://docs.github.com/en/actions/writing-workflows/choosing-where-your-workflow-runs/running-jobs-in-a-container) and packages and using it to run the rest of the workflow steps is preferable for a number of reasons:
1. Reusability: Image can be used for other purposes, including locally for validating submissions.
2. Modularity: It can still form the basis of a container that does run validation code if we decide this is necessary. It can even be used by hub admins as the basis for other containers.
3. The `validate_pr()` function has a number of arguments that allow for customisation of validation. Hiding it away in a container script in a different repo means folks can't access that or we would need to write extra code to pass all validation arguments as command line args which feels unnecessary work.
4. Equally if any custom validation functions require additional dependencies, they cannot be configured as [described in our vignette](https://hubverse-org.github.io/hubValidations/articles/deploying-custom-functions.html#additional-dependencies) using the current approach.

In addition before deploying any of this, I feel we need a clear system that will ensure images are up to date, ideally as automated as possible.

Give all this and how important validation is to hub functionality, I feel this issue requires a RFC and decision document before moving forward further.

**Comment by annakrystalli** [comment]
I also wonder whether we can use a rocker base image to avoid having to build R when we build our images https://rocker-project.org/

**Comment by annakrystalli** [comment]
Sorry! I think I confused what was going on when I looked at it last week but now that all the docs are in the repo, it's clear the way you've designed this can be used already as described above 🚀.

Given the step to install packages will no longer be required in the vanilla validation workflow, it would be good once we've decided to roll the above out in our validation workflows to include an example of how hub admins could amend their workflow to include additional dependencies. I imagine the vignette here in `hubValidations` might be a good place and perhaps the updated `hubverse-actions` documentation for the updated `validate-submission` action?

**Comment by annakrystalli** [comment]
One last question, should we be installing the released hubverse packages from r-universe instead of from GitHub? Would it be useful to maintain a dev version of the image with dev versions of the packages?

**Comment by zkamvar** [comment]
> should we be installing the released hubverse packages from r-universe instead of from GitHub? 

No. We are using {renv} to control the versioning, which expects an archival repository. Theoretically, [{renv} should fall back to GitHub if the package is from a universe repository](https://ropensci.org/blog/2022/01/06/runiverse-renv/), but I've found in https://github.com/rstudio/renv/issues/2068 that this is still not the case. 

> Would it be useful to maintain a dev version of the image with dev versions of the packages?

I think so.

**Comment by zkamvar** [comment]
> Give all this and how important validation is to hub functionality, I feel this issue requires a RFC and decision document before moving forward further.

I agree and I want to make it clear that my exercise here was to explore the space and come up with a prototype/minimal viable product. There are a lot of different layers that can be optimized and I wanted to explore them sufficiently before proposing a project.

**Comment by zkamvar** [comment]
> include an example of how hub admins could amend their workflow to include additional dependencies.

That's a good point. I'll have to look at https://rstudio.github.io/renv/articles/docker.html#dynamically-provisioning-r-libraries-with-renv to make sure our implementation can be extended.

**Comment by annakrystalli** [comment]
I know this is still WIP but I think some of the functionality might well be better as official `hubValidations` functionality. I'm thinking specifically the `get_error_data()` which would also resolve https://github.com/hubverse-org/hubValidations/issues/16 

It could be that we add it to the print method but only deploy through an argument (e.g. `show_errors = TRUE`)?

**Comment by zkamvar** [comment]
> I know this is still WIP but I think some of the functionality is really useful and might be better as official `hubValidations` functionality. I'm thinking specifically the `get_error_data()` which would also resolve [#16](https://github.com/hubverse-org/hubValidations/issues/16)
> 
> It could be that we add it to the print method but only deploy through an argument (e.g. `show_errors = TRUE`)?

I can take a stabby-stab at it.

**Comment by elray1** [comment]
> I'm now realizing that the renv.lock is over-complicating things a bit. The docker container could be built from a simple docker file.

In past investigations essentially independently replicated by @matthewcornell and me, we've landed on building R-based containers using renv in order to get the installation of specific R package versions that renv provides, and which we've found somewhat difficult to achieve without renv.  @matthewcornell has written up the procedures we've used here: https://github.com/reichlab/operational-models/blob/main/flu_ar2/..%2FREADME.md#requirementstxt-and-renvlock-details. I (and I'm sure Matt) would be interested to hear if you're envisioning something else specific to deal with this?

**Comment by zkamvar** [comment]
> we've landed on building R-based containers using renv in order to get the installation of specific R package versions that renv provides

I guess it really depends on what the container is for. If it's a drop-in replacement for our current GitHub workflow setup, then a simple Dockerfile that generates a container at release time is sufficient.

If, however, we would need this for future reproducibility, then the renv.lock file is an important element because we would be able to re-build the container from an old renv.lock.

I think for a container that should mirror the release and development versions of the packages , using a renv.lock might be a bit too prescriptive because we rebuild the containers every time we do a release, containing a snapshot of the R package environment at the time of release, which better reflects the testing environment and it reduces the burden of maintaining and synchronizing a renv.lock file across the dependent package containers (e.g. the renv.lock for hubUtils will serve as the base renv.lock for hubValidations).

**Comment by zkamvar** [comment]
That all being said, we might be able to have the Dockerfile even generate the renv.lock as an artifact when it's built.

**Comment by elray1** [comment]
makes sense. i guess there's a plan to distribute the built images via docker hub or similar, and that's what a user would effectively interact with?

**Comment by zkamvar** [comment]
> makes sense. i guess there's a plan to distribute the built images via docker hub or similar, and that's what a user would effectively interact with?

Yes. The plan is to use GitHub Workflows to build and deploy the containers to GitHub's container registry (ghcr.io). This is how https://github.com/hubverse-org/hub-dash-site-deployer and https://github.com/hubverse-org/hubPredEvalsData-docker are deployed.

**Comment by zkamvar** [comment]
This was put on hold because it was not being actively worked on. I want to provide some notes about why it was stalled out (sorry if they are rambly, there are a lot of moving parts):

## Scope

The scope of work for this project is to operationalize hubValidations in a docker image so that validations can be run without requiring a full R setup.

Success looks like:

- [ ] a docker image is released on ghcr.io or docker hub that tracks the released version of hubValidations
- [ ] a docker image is released on ghcr.io or docker hub that tracks the development version of hubValidations
- [ ] docker image versions are reproducible and can be rebuilt with a `renv.lock` file
- [ ] docker images can be extended
- [ ] workflows that use the docker image are added to https://github.com/hubverse-org/hubverse-actions

## Previous work

This is previous work to operationalize docker images, all with different degrees of manual intervention

- https://github.com/reichlab/operational-models
- https://github.com/hubverse-org/test-docker-hubValidations
- https://github.com/hubverse-org/hubPredEvalsData-docker 
- https://github.com/hubverse-org/hubValidations/issues/57#issuecomment-2599127972

## Blockers

Many of the blockers for this are related to general questions about how to operationalize docker images. Specifically, there are questions about building, dependencies, and release tracking that need to be generally addressed before we can achieve the goals above

### Building container reproducibly

We do not yet have consensus on how to build images reproducibly. One of the challenges here is understanding the fact that R packages development is unlike other languages in that **packages are implicitly assumed to be forward-compatible**, so you are always testing against the latest versions of your dependencies (unlike python and other languages, where you can set an upper limit on versions of your dependencies).

Below are some methods that we try and that I've seen

#### Building a container from a renv.lock file generated once
   
This is the workflow that is used in https://github.com/reichlab/operational-models, https://github.com/hubverse-org/test-docker-hubValidations, and https://github.com/hubverse-org/hubPredEvalsData-docker 

- pros
  - reproducible workflow
- cons
  - depends on the state of the user's machine at the time the image is created
  - unless the dependencies are explicitly updated in the lockfile, the image may return different results 

#### Building a container from a renv.lock file generated at build time

This is the scope of [the {pracpac} package](https://signaturescience.github.io/pracpac/)

- pros
  - reproducible workflow
  - image dependencies follow traditional workflow
- cons
  - depends on state of user's machine each time
  - [in {pracpac} system dependencies are not currently enumerated](https://github.com/signaturescience/pracpac/issues/105)
  - development dependencies (from `Description/Remotes:`) may sneak into package unless we are careful

#### Building a container and generating a renv.lock file _inside_ the container

This is the scope of [zkamvar/hubUtils/docker/release/Dockerfile](https://github.com/zkamvar/hubUtils/blob/9724a5488ab5d91a73b6e7ce629eaf8c6e723754/docker/release/Dockerfile#L35-L45).

- pros
  - same as pracpac
- cons
  - same as pracpac


### Dependencies vs image size

The easiest way to set up an image is to use `FROM rocker/verse:<tag>`, but that easily causes the built image to be somewhere between 5 and 6GB (because this bundles RStudio and publishing utilities as well)

A good compromise is to use `FROM rocker/r-ver:<tag>`, where the resulting image is under 2GB, but you don't have to manually set up R (though you do have to provision system dependencies)

#### System dependencies

System dependencies for hubValidations include libcurl and libxml. These need to be explicitly installed.

#### R dependencies

It's a bit of a challenge because the rocker project [tracks both R and Ubuntu dependencies](https://github.com/rocker-org/rocker-versioned2/wiki/Versions#rules) and **pins package releases to the last date that version of R was the default**. The safest way to build reproducibly is to build against the current version of R so that you get the current version of packages and the current OS.

Using an older version of R can lead to undesirable consequences such as a dependency on a version of {arrow} that was only released after the release of the next R version, which causes an error of "package not found" (yes, this has happened).

### Tracking Releases (where the docker source files live)

We have one package-based docker image operationalized for the dashboards (https://github.com/hubverse-org/hubPredEvalsData-docker). In order to update this to use the correct version of hubPredEvalsData, someone has to [manually run the update.sh script](https://github.com/hubverse-org/hubPredEvalsData-docker?tab=readme-ov-file#updating) (example in https://github.com/hubverse-org/hubPredEvalsData-docker/pull/3). 

All of our existing images run like this---they are external to the packages they wrap and require manual intervention to update.

A more pragmatic solution would be to have the Dockerfiles and associated side-acts live in a `docker/` directory of the package with corresponding workflows to rebuild the images on push and/or tag.
